File: TEAM B2 (Season 2)/Week 2/Musk/[스택_큐_주식가격].cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

// Stack / Queue를 이용한 방법이 아니라 일단 재껴놓고 다시 풀 예정
vector<int> solution(vector<int> prices) {
    vector<int> answer(prices.size());
    
    
    for(int i=0;i<prices.size();i++) {
        for(int j=i+1; j<prices.size(); j++) {
            answer[i]++;
            if(prices[i] > prices[j]) break;
        }
    }
    return answer;
}

File: TEAM B2 (Season 2)/Week 2/Ollie/[스택큐_주식가격].py
# 주식가격
# https://school.programmers.co.kr/learn/courses/30/lessons/42584
from typing import List

# [1, 2, 3, 2, 3] 	[4, 3, 1, 1, 0]
def solution(prices: List[int]) -> List[int]:
    stack = []
    answer = [0] * len(prices)
    for index, price in enumerate(prices):
        while stack and stack[-1][1] > price:
            pointer, _ = stack.pop()
            answer[pointer] = index - pointer
        stack.append((index, price))
    for index, _ in stack:
        answer[index] = len(prices) - index - 1
    return answer





# def solution(prices):
#     s = []
#     answer = [0] * len(prices)
#     for idx, price in enumerate(prices[:-1]):
#         while s and s[-1][1] > price:
#             p, _ = s.pop()
#             answer[p] = idx - p
#         s.append([idx, price])
#     for idx, _ in s:
#         answer[idx] = len(prices) - 1 - idx
#     return answer